File: user-service/src/main/java/a/gleb/user_service/service/OutboxMessageService.java
package a.gleb.user_service.service;

import a.gleb.user_service.db.repository.OutboxMessageRepository;
import jakarta.transaction.Transactional;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cloud.stream.function.StreamBridge;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class OutboxMessageService {

    private final StreamBridge streamBridge;
    private final OutboxMessageRepository outboxMessageRepository;

    /**
     * Send message to queue, and update message status.
     */
    @Transactional
    public void proceed() {
        outboxMessageRepository.findAllBySent(false)
                .forEach(it -> {
                    try {
                        streamBridge.send(it.getBindingName(), it.getMessage());
                        it.setSent(Boolean.TRUE);
                        outboxMessageRepository.save(it);
                    } catch (Exception e) {
                        log.error("send outbox message error", e);
                    }
                });
    }
}

File: main-service/src/main/kotlin/a/gleb/main_service/service/OutboxMessageService.kt
package a.gleb.main_service.service

import a.gleb.main_service.db.repository.OutboxMessageEntityRepository
import mu.KotlinLogging
import org.springframework.cloud.stream.function.StreamBridge
import org.springframework.stereotype.Service

private val logger = KotlinLogging.logger {}

@Service
class OutboxMessageService(
    private val outboxMessageRepository: OutboxMessageEntityRepository,
    private val streamBridge: StreamBridge
) {

    /**
     * Method for sending [a.gleb.main_service.db.entity.OutboxMessageEntity] to Kafka topic.
     */
    suspend fun processOutboxMessage() {
        outboxMessageRepository.findOutboxMessageEntityBySent(false)
            .forEach { outboxMessageEntity ->
                logger.info { "Processing outbox message: ${outboxMessageEntity.id}" }
                streamBridge.send(outboxMessageEntity.bindingName, outboxMessageEntity)
                outboxMessageEntity.sent = true
                outboxMessageRepository.save(outboxMessageEntity)
            }
    }
}